### ISSUE TITLE
Relationship Data - missing or by design?

### ISSUE THREAD

**Issue by mattgi** [json-api-dotnet/JsonApiDotNetCore#218]
Should I always see relationship data, regardless of link config or include query?

By this, I mean `/items` currently produces:

```json
{
    "data": [
        {
            "type": "items",
            "id": "item-de4e78d3-b2ec-48dd-8dd7-6088881a2bfe",
            "attributes": {
                "type": 0,
                "title": "New York",
                "cover": null,
                "slug": null,
                "version": 1
            },
            "relationships": {
                "owner": {
                    "links": {
                        "self": "/items/item-de4e78d3-b2ec-48dd-8dd7-6088881a2bfe/relationships/owner",
                        "related": "/items/item-de4e78d3-b2ec-48dd-8dd7-6088881a2bfe/owner"
                    },
                "createdBy": {},
                "updatedBy": {},
                "deletedBy": {}
                }
            }
        }
    ]
}
```

If I include `owner` (`/items?include=owner`):
```json
{
    "data": [
        {
            "type": "items",
            "id": "item-de4e78d3-b2ec-48dd-8dd7-6088881a2bfe",
            "attributes": {
                "type": 0,
                "title": "New York",
                "cover": null,
                "slug": null,
                "version": 1
            },
            "relationships": {
                "owner": {
                    "links": {
                        "self": "/items/item-de4e78d3-b2ec-48dd-8dd7-6088881a2bfe/relationships/owner",
                        "related": "/items/item-de4e78d3-b2ec-48dd-8dd7-6088881a2bfe/owner"
                    },
                    "data": {
                        "type": "users",
                        "id": "user-b607e7dc-3a9d-4360-8495-dd37c4c99de8"
                    }
                },
                "createdBy": {},
                "updatedBy": {},
                "deletedBy": {}
            }
        }
    ],
    "included": [
        {
            "type": "users",
            "id": "user-b607e7dc-3a9d-4360-8495-dd37c4c99de8",
            "attributes": {
                "email": "poo@shorthand.com",
                "firstName": "Matthew",
                "lastName": "Gint2y",
                "slug": "slugger",
                "version": 1
            },
            "relationships": {
                "createdBy": {},
                "updatedBy": {},
                "deletedBy": {}
            }
        }
    ]
}
```

Even when I am not asking for `includes` (or if `links` are turned off), I was expecting the relationships to always have data:

```json
"relationships": {
                "owner": {
                    "data": {
                        "type": "users",
                        "id": "user-b607e7dc-3a9d-4360-8495-dd37c4c99de8"
                    }
                },
```

Is this expected behavior or am I potentially missing something on my model?

```csharp
      [HasOne("owner")]
      public virtual User Owner { get; set; }

      public string OwnerId { get; set; }
```

Based on http://jsonapi.org/format/#document-resource-objects and their `{ article }` snippet, it suggests data should be there.. I must be missing something somewhere?

**Comment by jaredcnance** [comment]
I think this is an oversight. In the case of a dependent HasOne relationship we can include this easily and I think maybe we should. However, on the independent side of a HasOne (in the case of a 1:(0-1) relationship) we will not include this because it would necessitate a SQL JOIN and would be unexpected behavior.

**Comment by mattgi** [comment]
Cool. 

If the client has a local store, it may not need to re-fetch the relationship if it knows the `id` and `type` of the related item so this makes the most sense to me.

Is it a quick fix (e.g. could I PR without having to read through the entire library)?

**Comment by jaredcnance** [comment]
I think we need to do some setup when we build the ContextGraph so we don't do any extra request-time reflection. That would involve modifying the ContextGraphBuilder to set some HasOneAttribute.IsDependent and maybe even a pointer to the RelationshipId PropertyInfo. Then the DocumentBuilder will need to pull out the value. With tests, I would expect it to take me a couple hours. I'd also be happy to work with you if you want to take a stab at it.

**Comment by mattgi** [comment]
Ha, I just deleted my comment.

I was going to just try messing with the [DocumentBuilder](https://github.com/json-api-dotnet/JsonApiDotNetCore/blob/master/src/JsonApiDotNetCore/Builders/DocumentBuilder.cs#L157) 

Will start to get me used to the code if nothing else.

**Comment by mattgi** [comment]
A very quick hack was to:

1. copy the source `DocumentBuilder` into my local codebase,
2. Inject the copy in startup via `services.AddScoped<IDocumentBuilder, JsonBuilder>();`
3. Make this hack to `AddRelationships`:

```
private void AddRelationships(DocumentData data, ContextEntity contextEntity, IIdentifiable entity) {
...
  if (RelationshipIsIncluded(r.PublicRelationshipName)) {
...
  } else {
    relationshipData.SingleData = GetSimpleRelationship(entity, r.InternalRelationshipName);
  }
...
```

4. Add the corresponding method:
```
    private Dictionary<string, string> GetSimpleRelationship(object entity, string name) {
      var entityType = entity.GetType();
      var contextEntity = _contextGraph.GetContextEntity(entityType);
      var property = contextEntity.Relationships.First(f => f.InternalRelationshipName == name);
      if (property == null) {
        return null;
      }

      var propertyInfo = entityType.GetProperty(string.Concat(name, "Id"));
      if (propertyInfo == null) {
        return null;
      }

      var id = (string) (propertyInfo.GetValue(entity, null));
      if (id == null) {
        return null;
      }

      return new Dictionary<string, string> {
        {
          "type",
          property.Type.Name
        },
        { "id", id }
      };
    }
```

----

It assumes that when you have a `relationship` named `Something` on the entity that there will be a prop named `SomethingId` that stores the PK for that related entity (and that it's a string).

Very hacky, and I probably need to consider more complex models if I run in to them, but it quickly unblocked the clientside dev.

FWIW, I started to go down the rabbit hole of ContextGraph changes but quickly came back up for air before I sunk a day on it. :)

Thanks for pointing me in the right direction for now.

**Comment by crfloyd** [comment]
I currently have the need to include "data" (without attributes) for each relationship by default. From what I could tell, the only thing preventing this is the check in DocumentBuilder.cs in the AddRelationships() method: 
``if (RelationshipIsIncluded(r.PublicRelationshipName))``
removing this check just includes the data for the relationship by default. I am using a document database and not the default EntityFramework style so I do not know if there are any implications to doing this outside my use-case. It would be nice to have the ability to at least configure the framework to all for the default inclusion of data for resources.

**Comment by jaredcnance** [comment]
The only challenge I see is this requirement from the spec (let me know if I'm just missing something here):

> If an endpoint supports the include parameter and a client supplies it, the server MUST NOT include unrequested resource objects in the included section of the compound document.

Without that check, I think there are conditions under which we would no longer comply with the specification. As a framework we have a strict constraint of complying with the spec by default. However, I 100% believe we should make it simple for applications to deviate as needed.

A possible option is to make `RelationshipIsIncluded` protected so you can simply `protected override bool RelationshipIsIncluded(string relationshipName) => true;` but I'm not sure that's the best solution.

**Comment by crfloyd** [comment]
But isn't that referring to the "included" section of the document response? This is different from the inclusion of the "data" field in the "relationships" section. Below is an example response showing a response with included data. This response returns the included information requested in the "included" section but still includes the "data" at the relationship level. This would always be present regardless of what included information was requested. 
![image](https://user-images.githubusercontent.com/5572685/39708808-c19ae734-51dd-11e8-9ba9-d1d582299565.png)

**Comment by jaredcnance** [comment]
Ah, yes sorry my mistake. You’re correct.

**Comment by wurmrobert** [comment]
Hy.

I'm using Version 2.1.0 and would like to receive the relationships data object an every GET request.  Is there any possibility to get only the { type, id } relationship infos without the included object?

some example:

`{
  "data": {
    "type": "photos",
    "attributes": {
      "title": "Ember Hamster",
      "src": "http://example.com/images/productivity.png"
    },
    "relationships": {
      "photographer": {
        "data": { "type": "people", "id": "9" }
      }
    }
  }
}`

**Comment by jaredcnance** [comment]
@wurmrobert please try the latest version (2.2.4) that fixes this issue. If that does not work for you, let me know.
https://github.com/json-api-dotnet/JsonApiDotNetCore/releases/tag/v2.2.4

**Comment by jaredcnance** [comment]
@wurmrobert this is working for me running the example app. And no this is default behavior. As previously mentioned, we will not perform a join to get dependent relationship side of a 1:(0-*) relationship from the independent side. Can you answer a few questions to help me understand your setup:

- Can you share your models? 
- Can you share your HTTP request?
- Have you made any customizations?
- Can you verify that your object actually has an independent relationship? For example, if `Article` has one `Author` through `Article.AuthorId`, make sure the foreign key value is not null.

![image](https://user-images.githubusercontent.com/13851721/41196073-3d05e4e8-6bfe-11e8-8145-5b375ab06f2b.png)

**Comment by wurmrobert** [comment]
@jaredcnance thanks a lot for your help!

my models: 

```
public class Customer: Identifiable {
        
        [Attr("first_name")]
        public string FirstName { get; set; }

        [HasMany("devices")]
        public virtual List<Device> Devices { get; set; }
}
```

```
public class Device: Identifiable {
        [Attr("mac_address")]
        public string MacAddress { get; set; }

        [HasOne("customer")]
        public virtual Customer Customer { get; set; }
}
```

`curl -X GET "http://localhost:5000/api/v1/devices"`
<img width="1006" alt="screen shot 2018-06-10 at 00 09 16" src="https://user-images.githubusercontent.com/9077841/41196540-b6117660-6c42-11e8-883c-327a9c480a15.png">

With Postico i can see that device has a relation to a customer:
<img width="701" alt="screen shot 2018-06-10 at 00 09 59" src="https://user-images.githubusercontent.com/9077841/41196541-ba992b88-6c42-11e8-80cd-5873d5d71303.png">

**Comment by jaredcnance** [comment]
`Device` needs a `CustomerId` property exposed on your model.

```c#
public class Device: Identifiable {
  [Attr("mac_address")]
  public string MacAddress { get; set; }

  [HasOne("customer")]
  public virtual Customer Customer { get; set; }
  public virtual int CustomerId { get; set; } // we need EF to populate this property
}

```

**Comment by jaredcnance** [comment]
I am unable to reproduce this behavior (we also have tests that [run this request](https://github.com/json-api-dotnet/JsonApiDotNetCore/blob/a1832f764a9861c06ed919d75b8a9a80fa1f525e/test/JsonApiDotNetCoreExampleTests/Acceptance/Spec/CreatingDataTests.cs#L58-L75). A few other things you might check:

- Are there any interesting logs produced when this request is executed?
- Can you share your Startup code where you add JADNC services?
- Is the a `DbSet<Customer> Customers { get; set; }` property on your `DbContext`?
- Have you done any special mapping of your `DbContext` (using the `ModelBuilder`) ?
- Can you debug your app and see what the incoming `Device` object looks like in the controller? Has the `Device.CustomerId` property actually been set?

**Comment by wurmrobert** [comment]
Really strange.

- No there aren´t any interesting logs
- I don´t do any special mapping with `ModelBuilder`
- The `Device.CustomerId` on the incoming device is not set.

I have put the source to a public repo: https://github.com/wurmrobert/jsonapi-graphql-example

The startup class:

`
namespace backend {
    public class Startup {
        public Startup (IConfiguration configuration) {
            Configuration = configuration;
        }

        public IConfiguration Configuration { get; }

        public void ConfigureServices (IServiceCollection services) {
            ....
            var connectionString = Configuration["DbContextSettings:ConnectionString"];
            services.AddEntityFrameworkNpgsql ()
                .AddDbContext<MasterdataContext> (options => options.UseNpgsql (connectionString));

            // add jsonapi dotnet core
            services.AddJsonApi<MasterdataContext> (
                opt => opt.Namespace = "api/v1"
            );
            ...
            services.AddMvc ();
        }

        public void Configure (IApplicationBuilder app, IHostingEnvironment env) {
            if (env.IsDevelopment ()) app.UseDeveloperExceptionPage ();
            ...
            app.UseJsonApi ();
            ...
        }
    }
}
`

**Comment by wurmrobert** [comment]
Hy @jaredcnance.

After setting customerId as optional:

`public int? CustomerId { get; set; }`

everything works as expected. Thank you.
